actionpack/lib/action_controller/metal/live.rb

```diff
@@ -141,7 +141,7 @@ def initialize(response)
         @cv = new_cond
         @aborted = false
         @ignore_disconnect = false
-        super(response, SizedQueue.new(10))
+        super(response, Rails.env.test? ? Queue.new : SizedQueue.new(10))
```

Comment:
It's a layering violation for us to be looking at `Rails.env` here... and within the framework there's nothing special about the name `test` anyway.

I think it could be worth looking at option 3, though "why a SizedQueue at all?" is a good question. Does checking `git blame` for the commit that introduced it reveal anything interesting?

Reply:
The change from an infinite queue to a finite one was made in this commit: https://github.com/rails/rails/commit/37a764b8e0af585b3475a54566816a2ffc52401b.

The commit message says "use a sized buffer to prevent the queue being too large". Do you have more context @tenderlove? Thank you!

Reply:
I just asked myself why just to change the behaviour just for the test environment. That makes no sense and it is overall dangerous. For example I am faceing the problem in a production environment, see https://github.com/rails/rails/issues/31200/issues/30006. It crashes the whole app. If you change this behaviour only for testing, the app will still break in production and worse you would not notice it while testing.

But another thing that makes me suspicious is that   https://github.com/rails/rails/commit/37a764b8e0af585b3475a54566816a2ffc52401b is dating back to 2012. My app was running fine for years with this code until migrating to rails 5. So this can not cause the deadlock alone. Something else must be changed that leads to this issue.

Reply:
I ran into this same problem.  It looks like this is only needed because of the [ActionController::Live monkey patch in action_controller/test_case.rb](https://github.com/rails/rails/blob/v6.1.3/actionpack/lib/action_controller/test_case.rb#L18-L27) so it seems like it would make sense to have a corresponding monkey patch for this queue.  They are coupled together, since that existing patch makes the response synchronous, so the test can't consume the queue as it is written to in order to avoid it filling up and deadlocking.

Reply:
Opened https://github.com/rails/rails/pull/41609 which uses an attribute on the class to specify the queue length, so it can be overridden in action_controller/test_case.rb using just an attribute assignment.  Hopefully that is clean enough to be accepted.